---
title: "A tour of MicrobiomeExplorer"
author: "Sabbir Hossain"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: 
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{A tour of MicrobiomeExplorer}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(fig.align = "center", 
               out.width = "90%",
               fig.width = 6, fig.height = 5.5,
               dev.args=list(pointsize=10),
               collapse = TRUE,
               warning = FALSE)
set.seed(1) # for exact reproducibility
```

## Introduction

`MicrobiomeExplorer` is an R package designed to facilitate the analysis and 
visualization of microbiome datasets. It includes functions to calculate 
statistical metrics, normalize data, and create comprehensive heatmaps for 
exploratory data analysis. 
**This document provides a guide to using the MicrobiomeExplorer package (version 0.1.0)**.

To install **MicrobiomeExplorer**, run the following commands:

You must first install its dependencies. Notably, phyloseq and ComplexHeatmap 
which are available through Bioconductor, not CRAN, and require Bioconductor's 
installation methods.

Please run the following commands to install the current version of the 
dependency packages:


To install the current version of `MicrobiomeExplorer` from GitHub, use:
```r
# Install devtools if not already installed
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
    
# Install MicrobiomeExplorer from GitHub
devtools::install_github("itsSabbir/MicrobiomeExplorer", build_vignettes = TRUE)

# Load MicrobiomeExplorer
library(MicrobiomeExplorer)

```

To list all functions available in the package:
``` r
ls("package:MicrobiomeExplorer")
```

To list all datasets available in the package:
``` r
data(package = "MicrobiomeExplorer")
```

<br>

## Using MicrobiomeExplorer

### Functions in the Package

`MicrobiomeExplorer` offers a suite of functions that facilitate the exploration and analysis of microbiome data. Here is a list of the main functions along with a brief description:

- `calculateStats`: Calculates extended statistics (mean, median, standard deviation, variance, range, and interquartile range) for each feature in a microbiome dataset, providing a comprehensive statistical overview essential for preliminary data analysis.

- `plotMicrobiomeHeatmap`: Creates a comprehensive heatmap for microbiome data analysis, with support for normalization, various clustering methods, and enhanced visualization options using the `ComplexHeatmap` package.

- `advancedRarefactionPlot`: Generates rarefaction curves for microbiome count data, useful for comparing species richness among samples with customizable sampling and plot aesthetics.

- `calculateAlphaDiversity`: Computes various alpha diversity indices (like Shannon, Simpson, Chao1, and ACE) for each sample in a microbiome dataset, facilitating in-depth ecological and diversity studies.

- `addData`: Adds new data to a MicrobiomeData object, ensuring class compatibility and handling overlapping samples with advanced merging strategies. Supports rRNA16S, Metagenomic, or Metatranscriptomic data types, with specific validation for each..

- `mergeData`: Helper function to merge data matrices within MicrobiomeData objects, offering advanced merging options for overlapping data (like averaging or prioritizing new/existing data).

- `removeData`: Removes specified data (rRNA16S, Metagenomic, or Metatranscriptomic) from a MicrobiomeData object, simplifying data management and manipulation.

- `updateSampleInfo`: Updates the sample information in a MicrobiomeData object, ensuring consistency and integrity of new sample data with existing datasets.

- `validateSampleInfo`: Validates new sample information against existing data in MicrobiomeData objects, checking for consistency in sample counts, uniqueness, and matching identifiers.

- `validate16SRNAData`: Checks 16S rRNA gene sequencing data for structure, quality, and content, including numeric values, absence of negative values, minimum column and non-zero entry requirements, and no missing values.

- `validateMetagenomicData`: Validates metagenomic data for structure and quality, ensuring numeric values, no negative or extremely high values, and adequate dimensions.

- `validateMetatranscriptomicData`: Ensures metatranscriptomic data integrity by checking format, numeric values, non-negativity, potential outliers, and sequencing depth consistency.

- `MicrobiomeData S4 Class`: Defines an S4 class for managing microbiome data, including types such as 16S rRNA gene sequencing, metagenomic, and metatranscriptomic data, along with sample information.

- `createMicrobiomeDataObject`: Initializes a MicrobiomeData object with provided data, performing validation checks on the input data.

- `getData`: Extracts specified data type ('rRNA16S', 'Metagenomic', 'Metatranscriptomic', 'SampleInfo') from a MicrobiomeData object.

- `updateData`: Updates specific data within a MicrobiomeData object, allowing modification of existing data slots.

- `performDifferentialExpression`: Conducts differential expression analysis on microbiome data using either DESeq2 or EdgeR methods. It filters data based on count thresholds, normalizes counts, and returns results of the statistical analysis.

- `runMicrobiomeExplorerApp`: Initiates a Shiny application for the MicrobiomeExplorer package. This app provides an interactive interface for visualizing, analyzing, and exploring microbiome data. It supports various functionalities like plot generation and statistical analysis. The function launches the app in the user's default web browser and is located in the `./inst/shiny-scripts` directory of the MicrobiomeExplorer package.

### Data Preparation and Formatting

Before analyzing your microbiome data, ensure it is in the correct format: 
a matrix or dataframe with rows representing samples and columns representing 
taxa.

Run the R following code block to generate some data called microbiome_example 
to be used in the vignettes:

```{r}
# Load in the data for the function tutorials and walkthroughs.

# Microbiome Example Dataset with Decimal Numbers
data("microbiome_example")
# Microbiome Example Dataset with Whole Numbers
data("microbiomeExample")
```


### Calculating Statistics

Use the `calculate_stats` function to compute extended statistics for your dataset:

```{r}
# Calculate statistics
stats <- MicrobiomeExplorer::calculate_stats(microbiome_example)
stats
```

### Visualizing Data

Visualize your microbiome data with a heatmap using the `plot_microbiome_heatmap` function:

``` r
# Generate a heatmap
heatmap_result <- plot_microbiome_heatmap(microbiome_example, normalize = TRUE)
print(heatmap_result)
```

<br>

## Conclusion

The `MicrobiomeExplorer` package provides an easy-to-use set of tools for 
microbiome data analysis, allowing researchers to quickly calculate key 
statistics and visualize their data. For any further details, consult the 
function-specific help pages.

<br>

## Session Information

It's always a good practice to include the session information in vignettes:

```{r, message=FALSE, tidy=TRUE}
sessionInfo("MicrobiomeExplorer")
```
